### ISSUE TITLE
Problem with save-2! / fipp.edn

### ISSUE THREAD

**Issue by jll63** [jepsen-io/jepsen#418]
My Jepsen (version 0.1.15) tests crash while writing the edn files. I reduced it to this example:

```clojure
(ns jepsen.fipp-test
  (:require [jepsen.core    :as jepsen]
            [jepsen.tests   :as tests]
            [jepsen.checker :as checker]
            [clojure.test   :as test]
            ))

(test/deftest crash-test
  (let [test (merge tests/noop-test
                    {
                     :start-time "now"
                     :concurrency 4
                     :active-histories (atom #{})
                     :name "fipp-test"
                     :checker    (checker/total-queue)
                     })
        ]
    (jepsen/run-case! test)
    (jepsen/analyze! test)))
```
I get this stack trace, similar to what I get when running the real tests:
```
ERROR in (crash-test) (FutureTask.java:122)
Uncaught exception, not in assertion.
expected: nil
  actual: java.util.concurrent.ExecutionException: java.lang.NullPointerException
 at java.util.concurrent.FutureTask.report (FutureTask.java:122)
    java.util.concurrent.FutureTask.get (FutureTask.java:192)
    clojure.core$deref_future.invokeStatic (core.clj:2300)
    clojure.core$future_call$reify__8439.deref (core.clj:6974)
    clojure.core$deref.invokeStatic (core.clj:2320)
    clojure.core$deref.invoke (core.clj:2306)
    clojure.core$map$fn__5851.invoke (core.clj:2753)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:51)
    clojure.lang.RT.seq (RT.java:531)
    clojure.core$seq__5387.invokeStatic (core.clj:137)
    clojure.core$dorun.invokeStatic (core.clj:3133)
    clojure.core$dorun.invoke (core.clj:3133)
    jepsen.store$save_2_BANG_.invokeStatic (store.clj:374)
    jepsen.store$save_2_BANG_.invoke (store.clj:370)
    jepsen.core$analyze_BANG_.invokeStatic (core.clj:467)
    jepsen.core$analyze_BANG_.invoke (core.clj:451)
    jepsen.fipp_test$fn__566.invokeStatic (fipp_test.clj:19)
    jepsen.fipp_test/fn (fipp_test.clj:8)
    clojure.test$test_var$fn__9707.invoke (test.clj:717)
    clojure.test$test_var.invokeStatic (test.clj:717)
    clojure.test$test_var.invoke (test.clj:708)
    clojure.test$test_vars$fn__9733$fn__9738.invoke (test.clj:735)
    clojure.test$default_fixture.invokeStatic (test.clj:687)
    clojure.test$default_fixture.invoke (test.clj:683)
    clojure.test$test_vars$fn__9733.invoke (test.clj:735)
    clojure.test$default_fixture.invokeStatic (test.clj:687)
    clojure.test$default_fixture.invoke (test.clj:683)
    clojure.test$test_vars.invokeStatic (test.clj:731)
    clojure.test$test_all_vars.invokeStatic (test.clj:737)
    clojure.test$test_ns.invokeStatic (test.clj:758)
    clojure.test$test_ns.invoke (test.clj:743)
    user$eval224$fn__345.invoke (form-init4183486679217982891.clj:1)
    clojure.lang.AFn.applyToHelper (AFn.java:156)
    clojure.lang.AFn.applyTo (AFn.java:144)
    clojure.core$apply.invokeStatic (core.clj:667)
    clojure.core$apply.invoke (core.clj:660)
    leiningen.core.injected$compose_hooks$fn__154.doInvoke (form-init4183486679217982891.clj:1)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invokeStatic (core.clj:665)
    clojure.core$apply.invoke (core.clj:660)
    leiningen.core.injected$run_hooks.invokeStatic (form-init4183486679217982891.clj:1)
    leiningen.core.injected$run_hooks.invoke (form-init4183486679217982891.clj:1)
    leiningen.core.injected$prepare_for_hooks$fn__159$fn__160.doInvoke (form-init4183486679217982891.clj:1)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.lang.AFunction$1.doInvoke (AFunction.java:31)
    clojure.lang.RestFn.invoke (RestFn.java:408)
    clojure.core$map$fn__5851.invoke (core.clj:2755)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:51)
    clojure.lang.Cons.next (Cons.java:39)
    clojure.lang.RT.boundedLength (RT.java:1788)
    clojure.lang.RestFn.applyTo (RestFn.java:130)
    clojure.core$apply.invokeStatic (core.clj:667)
    clojure.test$run_tests.invokeStatic (test.clj:768)
    clojure.test$run_tests.doInvoke (test.clj:768)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invokeStatic (core.clj:665)
    clojure.core$apply.invoke (core.clj:660)
    user$eval224$fn__357$fn__410.invoke (form-init4183486679217982891.clj:1)
    user$eval224$fn__357$fn__358.invoke (form-init4183486679217982891.clj:1)
    user$eval224$fn__357.invoke (form-init4183486679217982891.clj:1)
    user$eval224.invokeStatic (form-init4183486679217982891.clj:1)
    user$eval224.invoke (form-init4183486679217982891.clj:1)
    clojure.lang.Compiler.eval (Compiler.java:7176)
    clojure.lang.Compiler.eval (Compiler.java:7166)
    clojure.lang.Compiler.load (Compiler.java:7635)
    clojure.lang.Compiler.loadFile (Compiler.java:7573)
    clojure.main$load_script.invokeStatic (main.clj:452)
    clojure.main$init_opt.invokeStatic (main.clj:454)
    clojure.main$init_opt.invoke (main.clj:454)
    clojure.main$initialize.invokeStatic (main.clj:485)
    clojure.main$null_opt.invokeStatic (main.clj:519)
    clojure.main$null_opt.invoke (main.clj:516)
    clojure.main$main.invokeStatic (main.clj:598)
    clojure.main$main.doInvoke (main.clj:561)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.lang.Var.applyTo (Var.java:705)
    clojure.main.main (main.java:37)
Caused by: java.lang.NullPointerException: null
 at multiset.core.MultiSet.iterator (core.clj:73)
    clojure.lang.RT.iter (RT.java:569)
    clojure.core$sequence.invokeStatic (core.clj:2663)
    clojure.core$sequence.invoke (core.clj:2647)
    fipp.edn$pretty_coll.invokeStatic (edn.cljc:15)
    fipp.edn$pretty_coll.invoke (edn.cljc:8)
    fipp.edn.EdnPrinter.visit_set (edn.cljc:65)
    fipp.visit$visit_STAR_.invokeStatic (visit.cljc:48)
    fipp.visit$visit_STAR_.invoke (visit.cljc:32)
    fipp.visit$visit.invokeStatic (visit.cljc:61)
    fipp.visit$visit.invoke (visit.cljc:58)
    fipp.edn.EdnPrinter/fn (edn.cljc:62)
    fipp.edn$pretty_coll$fn__1993.invoke (edn.cljc:12)
    clojure.core$map$fn__5847$fn__5848.invoke (core.clj:2742)
    clojure.lang.TransformerIterator.step (TransformerIterator.java:79)
    clojure.lang.TransformerIterator.hasNext (TransformerIterator.java:97)
    clojure.lang.RT$4.invoke (RT.java:518)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:51)
    clojure.lang.RT.seq (RT.java:531)
    clojure.core$seq__5387.invokeStatic (core.clj:137)
    clojure.core$map$fn__5851.invoke (core.clj:2746)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:51)
    clojure.lang.RT.seq (RT.java:531)
    clojure.core$seq__5387.invokeStatic (core.clj:137)
    clojure.core$apply.invokeStatic (core.clj:660)
    clojure.core$mapcat.invokeStatic (core.clj:2783)
    clojure.core$mapcat.doInvoke (core.clj:2783)
    clojure.lang.RestFn.invoke (RestFn.java:423)
    fipp.engine$serialize.invokeStatic (engine.cljc:15)
    fipp.engine$serialize.invoke (engine.cljc:12)
    clojure.core$map$fn__5851.invoke (core.clj:2753)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:51)
    clojure.lang.RT.seq (RT.java:531)
    clojure.core$seq__5387.invokeStatic (core.clj:137)
    clojure.core$apply.invokeStatic (core.clj:660)
    clojure.core$mapcat.invokeStatic (core.clj:2783)
    clojure.core$mapcat.doInvoke (core.clj:2783)
    clojure.lang.RestFn.invoke (RestFn.java:423)
    fipp.engine$serialize.invokeStatic (engine.cljc:15)
    fipp.engine$serialize.invoke (engine.cljc:12)
    fipp.engine$fn__1953.invokeStatic (engine.cljc:63)
    fipp.engine/fn (engine.cljc:58)
    clojure.lang.MultiFn.invoke (MultiFn.java:229)
    fipp.engine$serialize.invokeStatic (engine.cljc:18)
    fipp.engine$serialize.invoke (engine.cljc:12)
    clojure.core$map$fn__5851.invoke (core.clj:2753)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:51)
    clojure.lang.RT.seq (RT.java:531)
    clojure.core$seq__5387.invokeStatic (core.clj:137)
    clojure.core$apply.invokeStatic (core.clj:660)
    clojure.core$mapcat.invokeStatic (core.clj:2783)
    clojure.core$mapcat.doInvoke (core.clj:2783)
    clojure.lang.RestFn.invoke (RestFn.java:423)
    fipp.engine$serialize.invokeStatic (engine.cljc:15)
    fipp.engine$serialize.invoke (engine.cljc:12)
    fipp.engine$fn__1938.invokeStatic (engine.cljc:48)
    fipp.engine/fn (engine.cljc:47)
    clojure.lang.MultiFn.invoke (MultiFn.java:229)
    fipp.engine$serialize.invokeStatic (engine.cljc:18)
    fipp.engine$serialize.invoke (engine.cljc:12)
    fipp.engine$pprint_document.invokeStatic (engine.cljc:236)
    fipp.engine$pprint_document.invoke (engine.cljc:234)
    fipp.edn$pprint.invokeStatic (edn.cljc:100)
    fipp.edn$pprint.invoke (edn.cljc:91)
    fipp.edn$pprint.invokeStatic (edn.cljc:92)
    fipp.edn$pprint.invoke (edn.cljc:91)
    jepsen.store$write_results_BANG_.invokeStatic (store.clj:334)
    jepsen.store$write_results_BANG_.invoke (store.clj:330)
    jepsen.store$save_2_BANG_$fn__3915.invoke (store.clj:374)
    clojure.core$binding_conveyor_fn$fn__5739.invoke (core.clj:2030)
    clojure.lang.AFn.call (AFn.java:18)
    java.util.concurrent.FutureTask.run (FutureTask.java:266)
    java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    java.lang.Thread.run (Thread.java:748)
```

When I hack `store.clj` to use `clojure.pprint` instead of `fipp.edn`, the problem goes away.

Commenting out the checker prevents the crash.

I tried upgrading `fipp.edn` to 0.6.19, it still crashes.

**Comment by aphyr** [comment]
This is patched in 0.1.16-SNAPSHOT, available on Clojars.

**Comment by jll63** [comment]
I tried it and I still get this error. I use a company mirror but I think I have the right version in my `.m2`, i.e. `jepsen-0.1.16-20190917.131108-1.jar`, sha1 is `1b7fd79f82e9c3716262b86ffd659b02bd1f49c3`.

**Comment by aphyr** [comment]
Huh! It's gonna be a few days before I have access to a computer where I can debug, but in the meantime, maybe check your lein deps :tree to look for instances of the multiset package? There was a bug in multiset that broke traversal of empty sets--possibly there's an old version hanging around in the tree somewhere.On Oct 3, 2019 10:41, Jean-Louis Leroy <notifications@github.com> wrote:I tried it and I still get this error. I use a company mirror but I think I have the right version in my .m2, i.e. jepsen-0.1.16-20190917.131108-1.jar, sha1 is 1b7fd79f82e9c3716262b86ffd659b02bd1f49c3.

**Comment by jll63** [comment]
No worry, I can get by with my hacked version for the time being.

I should point out that I am new to Clojure, lein, even Java. Still bluffing my way into this ;-)

I see:

```
   [knossos "0.3.4" :exclusions [[org.slf4j/slf4j-log4j12]]]
     [com.boundary/high-scale-lib "1.0.6"]
     [interval-metrics "1.0.0"]
     [org.clojars.achim/multiset "0.1.0"]
     [org.clojars.pallix/analemma "1.0.0" :exclusions [[org.clojure/clojure]]]
     [org.clojure/math.combinatorics "0.1.1"]
     [potemkin "0.4.3"]
       [riddley "0.1.12"]
   [metametadata/multiset "0.1.1"]
     [org.clojure/algo.generic "0.1.2"]
```
When I build the uberjar and look inside it, I see `metadata/multiset` version 0.1.1.

**Comment by aphyr** [comment]
Okay! It might also be stale state somewhere in the build--sometimes a `lein clean` does the trick. :)On Oct 3, 2019 11:07, Jean-Louis Leroy <notifications@github.com> wrote:No worry, I can get by with my hacked version for the time being.
I should point out that I am new to Clojure, lein, even Java. Still bluffing my way into this ;-)
I see:
   [knossos "0.3.4" :exclusions [[org.slf4j/slf4j-log4j12]]]
     [com.boundary/high-scale-lib "1.0.6"]
     [interval-metrics "1.0.0"]
     [org.clojars.achim/multiset "0.1.0"]
     [org.clojars.pallix/analemma "1.0.0" :exclusions [[org.clojure/clojure]]]
     [org.clojure/math.combinatorics "0.1.1"]
     [potemkin "0.4.3"]
       [riddley "0.1.12"]
   [metametadata/multiset "0.1.1"]
     [org.clojure/algo.generic "0.1.2"]

When I build the uberjar and look inside it, I see metadata/multiset version 0.1.1.

**Comment by jll63** [comment]
I tried `lein clean`, to no effect. 

Using an exclusion in my `project.clj` (`[jepsen "0.1.16-SNAPSHOT" :exclusions [org.clojars.achim/multiset]]`) I managed to get rid of the "older" multiset dep, according to `lein deps :tree`, but the test still crashes.

**Comment by jll63** [comment]
Fixed in `fbee1b4`
